homeassistant/components/mqtt/__init__.py

```diff
@@ -307,16 +324,16 @@ def async_setup(hass: HomeAssistantType, config: ConfigType):
         broker_config = yield from _async_setup_server(hass, config)
 
     if CONF_BROKER in conf:
-        broker = conf[CONF_BROKER]
-        port = conf[CONF_PORT]
-        username = conf.get(CONF_USERNAME)
-        password = conf.get(CONF_PASSWORD)
-        certificate = conf.get(CONF_CERTIFICATE)
-        client_key = conf.get(CONF_CLIENT_KEY)
-        client_cert = conf.get(CONF_CLIENT_CERT)
-        tls_insecure = conf.get(CONF_TLS_INSECURE)
-        protocol = conf[CONF_PROTOCOL]
-    elif broker_config:
+        broker = conf[CONF_BROKER]  # type: str
+        port = conf[CONF_PORT]  # type: int
+        username = conf.get(CONF_USERNAME)  # type: Optional[str]
+        password = conf.get(CONF_PASSWORD)  # type: Optional[str]
+        certificate = conf.get(CONF_CERTIFICATE)  # type: Optional[str]
+        client_key = conf.get(CONF_CLIENT_KEY)  # type: Optional[str]
+        client_cert = conf.get(CONF_CLIENT_CERT)  # type: Optional[str]
+        tls_insecure = conf.get(CONF_TLS_INSECURE)  # type: Optional[bool]
+        protocol = conf[CONF_PROTOCOL]  # type: str
```

Comment:
Sometimes we do `conf[...]`, sometimes `conf.get(...)`. I get why `conf.get(..., default)` can be useful for default values, but those aren't used here. What doe you prefer? (I would be happy to make those statements more consistent *in this file* 😉.)

Reply:
conf[...] will throw an exception
conf.get(...) will return None

quite different I would say ;)

Reply:
Oh sure, you're right! (I somehow though voluptuous would take care of this, but apparently not)

Reply:
I would prefer if we always would use `conf[key]` for required keys and `conf.get(key)` for optional keys.

Reply:
Advantage of raising `KeyError` is that we fail when the error happens, not somewhere down the line when a value is all of a sudden unexpectedly `None`